### PR TITLE
H1 FROST function, BIP034 hashing function

### PR DESCRIPTION
**Pull request by pdyraga**
~~Depends on #1. Let's merge #1 before we undraft/merge this one.~~

This changeset adds `h1` function. `h1` is the implementation of H1(m) function from [FROST] implemented in a way 
 compatible to how [BIP340] hashing functions are specified. The implementation is copied from @eth-r's prototype. I added tests and annotated relevant parts of the hash functions code with reference to papers. The `testutils` library code was borrowed from the `keep-network/keep-core` repository.

### CODE REVIEW

File: roast/hash.go [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+func h1(m []byte) *big.Int {
+	dst := concat(contextString, []byte("rho"))
+	return hashToScalar(dst, m)
+}
+
+// hashToScalar is the implementation of hash_to_field(msg, 1) from
+// [HASH-TO-CURVE] with m=1 as specified in section 6.5. FROST(secp256k1, SHA-256)
+// of [FROST]. The [BIP340] hash function is used as expand_message_xmd.
+func hashToScalar(tag, msg []byte) *big.Int {
+	// The hash_to_field function is specified in [HASH-TO-CURVE] in section
```
Comment (eth-r) [inline roast/hash.go]: We are not actually doing this when implementing BIP-340 which has a much simpler specification for hashing scalars.

File: roast/hash.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+
+import (
+	"crypto/sha256"
+	"math/big"
+)
+
+// The contextString as defined in section 6.5. FROST(secp256k1, SHA-256) of
+// [FROST] is "FROST-secp256k1-SHA256-v1". Since we do a BIP-340 specialized
+// version, we use "FROST-secp256k1-BIP340-v1" as the contextString.
+var contextString = []byte("FROST-secp256k1-SHA256-v1")
```
Comment (eth-r) [inline roast/hash.go]: Mismatch with comment above

File: roast/roast.go [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+//	Threshold Schnorr Signatures with FROST", Work in Progress, Internet-Draft,
+//	draft-irtf-cfrg-frost-15, 5 December 2023,
+//	<https://datatracker.ietf.org/doc/draft-irtf-cfrg-frost/15/>.
+//
+// [HASH-TO-CURVE]
+//
+//	Faz-Hernandez, A., Scott, S., Sullivan, N., Wahby, R. S., and C. A. Wood,
+//	"Hashing to Elliptic Curves", Work in Progress, Internet-Draft,
+//	draft-irtf-cfrg-hash-to-curve- 16, 15 June 2022,
+//	<https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve-16>.
```
Comment (eth-r) [inline roast/roast.go]: I think HASH-TO-CURVE is just superseded by the BIP-340 hashing specification.

**Review by eth-r — CHANGES_REQUESTED** [review]
A couple of errors regarding the comments and the context string. I don't think a BIP-340 actually uses HASH-TO-CURVE specifications for anything because it has its own, simpler hashing rules.

File: roast/hash.go (hunk 2 above)
Comment (pdyraga) [reply roast/hash.go]: Fixed in https://github.com/threshold-network/roast-go/pull/2/commits/7644d8031a9291e45314854d55566cc2402c0a56.

File: roast/hash.go (hunk 1 above)
Comment (pdyraga) [reply roast/hash.go]: 👍🏻 Please see https://github.com/threshold-network/roast-go/pull/2/commits/bac833c55b4a5950897ab2193bb854fc37594bb8.

File: roast/roast.go (hunk 3 above)
Comment (pdyraga) [reply roast/roast.go]: See https://github.com/threshold-network/roast-go/pull/2/commits/bac833c55b4a5950897ab2193bb854fc37594bb8 , please.

**Comment by pdyraga** [comment]
Note that this specification is specific to BIP340, in the next PR I am planning to abstract out an interface that would allow us to provide another hashing strategy.

**Review by eth-r — APPROVED** [review]

**Pull request merged**